### ISSUE TITLE
[Enhancement][OpenLDAP] - Auto-delete groups when all users have been detached to prevent exception

### ISSUE THREAD

**Issue by cheesegrits** [DirectoryTree/LdapRecord#539]
**Environment:**
 - OpenLDAP
 - PHP 8.2

**Describe the bug:**

I don't know if this is really an LdapRecord "bug", as it's due to an OpenLDAP 'groupOfUniqueNames' issue ... but trying to detach() the last member of a group throws an exception because LDAP groups have to have at least one member.  So my code winds up looking like ...

```
                    if ($group->members()->count() === 1) {
                        try {
                            $group->delete();
                        }
                        catch (\Exception $e)
                        {
                            Log::channel('ldap')->error('Error deleting group: ' . $e->getMessage());
                        }
                    }
                    else {
                        $group->members()->detach($ldapUser);
                    }
```

... with the only real option being to delete the group in LDAP.

Maybe this could be an option in the LdapRecord config, or in the OpenLDAP group model, like 'delete_group_if_no_members' or some such?

**Comment by stevebauman** [comment]
Hi there @cheesegrits!

I personally don't think this should be added as a configuration option. 

LDAP servers vary widely in their implementations, so this should be up to the developer working with their LDAP distro to implement the behaviour they would like.

**Comment by cheesegrits** [comment]
That's fair enough.  Thanks for the response.

**Comment by cheesegrits** [comment]
> @cheesegrits Intentional re-open?

Sorry, yes, I'd re-opened then got distracted and never posted.

I just ran in to the issue again, in a case which I think makes a configurable option worth considering.

When deactivating a user, I'm using $user->groups()->detachAll(), exactly like your "off-boarding" example.  But if that user happens to be the last one in a given group, that throws an LDAP exception.  So I have to then manually fetch all their groups and iterate, find the one they are the only member of, delete it, then either retry the detachAll() or just manually detach() each one.

This essentially makes detachAll() unusable, as I might as well just iterate and detach() in the first place.

After googling around for this issue, I found 15 years worth of people complaining about it, and only two options to solve it.  Apparently, the actual RFC 2256 for LDAP says that 'member' in groupOfNames is a 'must', not a 'may', meaning that the actual LDAP protocol says groupOfNames cannot be empty.  Obviously empty groups are a perfectly reasonable thing to want, so people either modify the objectClass for groupOfnames and move 'member' to 'may', but that's just not a maintainable solution.  Or they add a dummy user to every group.

You could obviously argue that this is an application level problem, and we (at the app level) should handle adding a dummy user to every group we create.  Just seems like this could be something LdapRecord could maybe provide an option for as a convenience, seeing as how it's a known issue that requires a workaround for a lot of people.  Just an optional property on the OpenLDAP\Group class like

protected $dummyUser = 'do_not_delete@nowhere.com';

... and if it's not NULL, LdapRecord adds that user to any group being created.

Obviously I could handle this myself, but maybe worth considering.

**Comment by stevebauman** [comment]
Hi @cheesegrits! Thanks for following up.

That sounds really painful to have to workaround. I agree. Let's add this in as an option on the included OpenLDAP models.

I think having a unique method on the relationship will be the way to go here. Maybe something like:

```php
$user->groups()->detachAllOrDelete();
```

Thoughts?

**Comment by cheesegrits** [comment]
Thanks for the fast response.

That would certainly work for me, as in my use case it doesn't matter if empty groups are deleted, as the consumer of my directory only cares about which groups users are members of, it doesn't care about the existence of groups themselves.  I can imagine some use cases where the dummy user workaround would be needed, if "empty" groups need to stay in LDAP, but that wouldn't be too difficult to implement at the app level.

How about a $group->members()->detachOrDelete($ldapUser) as well?  That would also clean up my code from the OP in the thread.

**Comment by cheesegrits** [comment]
Or, maybe just a simple switch, like 

`protected bool $deleteEmptyGroups = true;`

... which then informs any mechanism which deletes users from that group class.

**Comment by stevebauman** [comment]
> How about a $group->members()->detachOrDelete($ldapUser) as well? That would also clean up my code from the OP in the thread.

Yea I like this syntax, as it's explicit and easy to understand.

If you have priority on this then I'd definitely accept a PR, otherwise I should be able to have this done in a couple days 👍 

I appreciate your help and input with this! 🙏

**Comment by cheesegrits** [comment]
Not that urgent, as I'm catching the exceptions and handling them atm, it'd just be nice to not have to worry about it and have it handled in the method calls.

Thanks for your fast responses, it is much appreciated.  I've talked to our client about costing in a support sponsorship in the monthly maintenance package, and they are happy to do it given that LdapRecord is a critical component of the app.  Hopefully that will happen later this week.  If you get contacted by someone called Robbie, that'll be us.

**Comment by stevebauman** [comment]
Of course @cheesegrits! Always happy to help!

And that's awesome! I appreciate any and all support -- paid or otherwise. Thank you ❤️

**Comment by stevebauman** [comment]
Hi @cheesegrits!

I've just released [v2.20.0](https://github.com/DirectoryTree/LdapRecord/releases/tag/v2.20.0) with both query chunking and  the `HasMany::detachOrDeleteParent()` and `HasMany::detachAllOrDelete()` methods.

I've altered the API for chunking in isolation -- as it's now opt-in (to prevent other devs from breaking changes). You can execute chunks on their own isolated connection by supplying `true` to its named parameter:

```php
User::each(function ($object) {
    // Additional queries here...
}, isolate: true);
```

Let me know if you encounter any issues with the addition of the relation methods mentioned above! Thanks for your help with these features 🙏
